Update governance ledger model

andrew-lee-work opened the pull request:
Update the governance ledger model to the Rust implementation as of Feb 2024.

* Deviations from the implementation are noted in a comment.
* Some bug fixes are included.

andrew-lee-work committed:
Remove reward minting from disburse_neuron

As of Feb 2024 disburse_neuron no longer mints rewards.
Fix locks.

andrew-lee-work committed:
Fix max amount in Disburse_To_Neuron

Account for neuron fees in the allowed range of
amounts to disburse. The original code allowed
the value of (parent_neuron.cached_stake - parent_neuron.fees)
to become negative after the transfer.

andrew-lee-work committed:
Fix typo in comment

andrew-lee-work committed:
Add burning of fees to merge_neurons

andrew-lee-work committed:
Remove Merge_Maturity_Of_Neuron

andrew-lee-work committed:
Add note on deviations from impl

andrew-lee-work committed:
Governance-TLA: Update TLA translation

andrew-lee-work committed:
Clean up refs to Merge_Maturity_Of_Neuron_Process_Ids

andrew-lee-work committed:
Fix refs to neuron id in Merge_Neurons

andrew-lee-work committed:
Governance-TLA: Fix Disburse_Neuron handling of fees

Do not allow disburse amount to go below zero due to neuron fees.

Disburse_Neuron was disbursing more than the actual stake.
If the neuron has 0 cached stake and 1 fees, this
Disburse_Neuron would attempt to send -1 tokens.

oggy-dfin commented on tla/governance-ledger/Governance.tla:
```diff
@@ -649,13 +621,42 @@ process ( Merge_Neurons \in Merge_Neurons_Process_Ids )
+            };
+        };
+
+    MergeNeurons3:
```
The `goto MergeNeuron5` below should be `goto MergeNeuron6` now I believe?

oggy-dfin commented on tla/governance-ledger/Governance.tla:
```diff
@@ -664,13 +665,13 @@ process ( Merge_Neurons \in Merge_Neurons_Process_Ids )
             };
         };
-    MergeNeurons3:
+    MergeNeurons4:
```
I can't find the balance query stuff in the code any more, but I find that the code has also unfortunately become a lot more difficult to understand; could you point me to where this query happens?

andrew-lee-work replied:
The implementation does not query ledger, and IIRC the implementation did not previously do that either, especially after already doing the transfer. I think this is an intentional divergence from the implementation in the TLA model. The additional query seems to be added to catch a violation of an invariant.

andrew-lee-work committed:
Governance-TLA: Remove error in MergeNeurons

Remove setting of `error` in MergeNeurons. This was accidentally copied
from DisburseNeuron without proper context.

andrew-lee-work committed:
Governance-TLA: Fix goto in MergeNeurons

andrew-lee-work committed:
Governance-TLA: Remove outdated comment

andrew-lee-work committed:
Governance-TLA: Translate TLA

oggy-dfin commented on tla/governance-ledger/Governance.tla:
```diff
@@ -95,12 +95,29 @@ Apalache caveats to keep in mind:
 
+
+***********
+Deviations
```
Could you add the merge_neuron differences here?

andrew-lee-work replied:
done

oggy-dfin reviewed (approved):
Pre-approving, please add a note about merge neuron to the deviations

andrew-lee-work committed:
Governance-TLA: Note Merge_Neurons deviation

Add a comment about the deviation in Merge_Neurons that
existed previously to the model updates in 2024.
